doc/developers/cython.rst

```diff
@@ -0,0 +1,178 @@
+.. _cython:
+
+Cython Best Practices, Conventions and Knowledge
+================================================
+
+This documents:
+
+* technical aspects of Cython which aren't currently present in Cython's documentation
+* tips to develop Cython code in scikit-learn
+
+Cython Internals
+----------------
+
+Concrete method dispatch implementation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* In most cases, concrete method dispatch is performed **at runtime using virtual tables**
+  (i.e. static C-array of pointers to a given types of function, used at runtime to known which
+  function to call from an instance). This is costly because there is a double indirection.
+
+* In some cases, concrete method dispatch is performed **statically** (a function is called
+  directly by a pointer dereference). This is less costly, especially if there are rooms
+  for this function to be inline.
+
+Known Limitations
+^^^^^^^^^^^^^^^^^
+
+* No template for types exist in Cython, alternatives are:
+
+  * using fused-types
+  * using Tempita, a small preprocessing language to expand Cython source
+
+* Multiple inheritance is not possible for extension types
+* No concept similar to interfaces or traits exist in Cython
+* Currently it is impossible to inherit and override a method which has a fused type in its signature
+* Structs cannot be created using constructor syntax with `nogil` `issue and workaround <https://github.com/cython/cython/issues/1642>`_
+
+Performances improvements will come with `Cython 3.0 which is to be released soon in beta apparently <https://github.com/cython/cython/issues/4022#issuecomment-1340918590>`_
+
+Basics
+^^^^^^
+
+* ``nogil`` declaration: https://github.com/cython/cython/issues/2798#issuecomment-459971828
```

Comment:
Instead of linking to a github comment, I'd rather rewrite the full explanation here. And probably explain in more detail in which situation we want to release the gil, i.e. when the code is meant to be used in a multi-threaded context.